x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -0,0 +1,133 @@
+import org.elasticsearch.action.index.IndexAction;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.search.SearchAction;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.search.sort.SortOrder;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelConfig;
+import org.elasticsearch.xpack.core.ml.inference.persistence.InferenceIndexConstants;
+import org.elasticsearch.xpack.core.ml.job.messages.Messages;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class TrainedModelProvider {
+
+    private static final Logger logger = LogManager.getLogger(TrainedModelProvider.class);
+    private final Client client;
+    private final NamedXContentRegistry xContentRegistry;
+
+    public TrainedModelProvider(Client client, NamedXContentRegistry xContentRegistry) {
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    public void storeTrainedModel(TrainedModelConfig trainedModelConfig, ActionListener<Boolean> listener) {
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            XContentBuilder source = trainedModelConfig.toXContent(builder, ToXContent.EMPTY_PARAMS);
+
+            IndexRequest indexRequest = new IndexRequest(InferenceIndexConstants.LATEST_INDEX_NAME)
+                .opType(DocWriteRequest.OpType.CREATE)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .id(TrainedModelConfig.documentId(trainedModelConfig.getModelId(), trainedModelConfig.getModelVersion()))
+                .source(source);
+            executeAsyncWithOrigin(client, ML_ORIGIN, IndexAction.INSTANCE, indexRequest,
+                ActionListener.wrap(
+                    r -> listener.onResponse(true),
+                    e -> {
+                        logger.error(
+                            new ParameterizedMessage("[{}] failed to store trained model for inference", trainedModelConfig.getModelId()),
```

Comment:
We use `Messages.INFERENCE_FAILED_TO_STORE_MODE` in line 80. Should it be used here as well?

Reply:
seems to me it should be moved into the if-clause and log corresponding messages

Reply:
I consider the two messages to be different. I think potentially user facing errors should be sentences (or close to it). Logged messages should follow the prevailing format of `[<RESOURCE_ID>] <MESSAGE>`

Reply:
I updated the message to include the model version. But again, for logging, it is much easier to grep and visually parse logs if the resource ID is right at the front. This is the prevailing pattern in our ML plugin and the Transform Plugin.

Reply:
(nit) should not have piggy backed my comment: For me it would be clearer if logging corresponds to the `listener.onFailure` calls, so we have 2 different log messages for the 2 failure cases.